### PR TITLE
[Fix-11197] [Master] Dependency detection bug

### PR DESCRIPTION
**Pull request by brave-lee**
closed by #11197

### CODE REVIEW

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/utils/DependentExecute.java [hunk 1]
```diff
@@ -186,15 +233,12 @@ private DependResult getDependTaskResult(long taskCode, ProcessInstance processI
      * @return ProcessInstance
      */
     private ProcessInstance findLastProcessInterval(Long definitionCode, DateInterval dateInterval) {
-
-        ProcessInstance runningProcess = processService.findLastRunningProcess(definitionCode,
-                dateInterval.getStartTime(), dateInterval.getEndTime());
+        ProcessInstance runningProcess = processService.findLastRunningProcess(definitionCode, dateInterval.getStartTime(), dateInterval.getEndTime());
```
Comment (caishunfeng) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/utils/DependentExecute.java]: related pr: https://github.com/apache/dolphinscheduler/pull/11424

File: dolphinscheduler-dao/src/main/resources/org/apache/dolphinscheduler/dao/mapper/TaskInstanceMapper.xml [hunk 2]
```diff
@@ -246,4 +246,29 @@
...
+        select
+        <include refid="baseSql"/>
+        from t_ds_task_instance
+        where task_code=#{taskCode}
+        <if test="startTime!=null and endTime != null">
+            and start_time <![CDATA[ >= ]]> #{startTime} and start_time <![CDATA[ <= ]]> #{endTime}
+        </if>
+        order by end_time desc limit 1
+    </select>
+    <select id="queryLastTaskInstanceList" resultType="org.apache.dolphinscheduler.dao.entity.TaskInstance">
```
Comment (caishunfeng) [inline dolphinscheduler-dao/src/main/resources/org/apache/dolphinscheduler/dao/mapper/TaskInstanceMapper.xml:259]:
```suggestion
    <select id="queryTaskInstanceListByTime" resultType="org.apache.dolphinscheduler.dao.entity.TaskInstance">
```

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/utils/DependentExecute.java [hunk 3]
```diff
@@ -129,51 +140,87 @@ private DependResult calculateResultForTasks(DependentItem dependentItem,
...
     /**
      * depend type = depend_all
-     *
-     * @return
      */
-    private DependResult dependResultByProcessInstance(ProcessInstance processInstance) {
-        if (!processInstance.getState().isFinished()) {
-            return DependResult.WAITING;
-        }
+    private DependResult dependResultByProcessInstance(ProcessInstance processInstance, DateInterval dateInterval) {
```
Comment (caishunfeng) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/utils/DependentExecute.java:144]: Please add some comments.

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/utils/DependentExecute.java [hunk 4]
```diff
@@ -129,51 +140,87 @@ private DependResult calculateResultForTasks(DependentItem dependentItem,
...
         if (processInstance.getState().isSuccess()) {
+            List<ProcessTaskRelation> taskRelations = processService.findRelationByCode(processInstance.getProcessDefinitionCode(),
+                processInstance.getProcessDefinitionVersion());
+            if (!taskRelations.isEmpty()) {
+                List<TaskDefinitionLog> taskDefinitionLogs = processService.genTaskDefineList(taskRelations);
+                Map<Long, String> definiteTask = taskDefinitionLogs.stream().filter(log -> !log.getTaskType().equals(TaskConstants.TASK_TYPE_SUB_PROCESS)
```
Comment (caishunfeng) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/utils/DependentExecute.java:150]: Why filter these logic tasks?

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/utils/DependentExecute.java [hunk 5]
```diff
@@ -129,51 +140,87 @@ private DependResult calculateResultForTasks(DependentItem dependentItem,
...
+                                v = instance;
+                            } else {
+                                if (v.getId() < instance.getId()) {
+                                    v = instance;
+                                }
+                            }
+                            return v;
+                        });
+                        definiteTask.remove(instance.getTaskCode());
+                    }
```
Comment (caishunfeng) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/utils/DependentExecute.java:174]: What is the point of this part?

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/utils/DependentExecute.java [hunk 6]
```diff
@@ -129,51 +140,87 @@ private DependResult calculateResultForTasks(DependentItem dependentItem,
...
-     *
-     * @return
      */
-    private DependResult dependResultByProcessInstance(ProcessInstance processInstance) {
-        if (!processInstance.getState().isFinished()) {
-            return DependResult.WAITING;
-        }
+    private DependResult dependResultByProcessInstance(ProcessInstance processInstance, DateInterval dateInterval) {
         if (processInstance.getState().isSuccess()) {
+            List<ProcessTaskRelation> taskRelations = processService.findRelationByCode(processInstance.getProcessDefinitionCode(),
```
Comment (lenboo) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/utils/DependentExecute.java:146]: The workflow dependent only need check the process instance state.
There are scenarios in which only some tasks are run, resulting in a successful workflow.

**Review by lenboo — CHANGES_REQUESTED** [review]

**Comment by lordk911** [comment]
Any new development?

**Comment by lordk911** [comment]
@SbloodyS Why this PR hung up such a long time, I think dep bug is important to be repair. or there is any other PR to resolve this problem? thanks.

**Comment by brave-lee** [comment]
refer #12219,  I think it need new pr to solve, someone want to contribute ?

**Comment by lordk911** [comment]
> refer #12219, I think it need new pr to solve, someone want to contribute ?

#12219 has no code changed but only doc , anything more need to do for is PR?

**Pull request closed without merging**